design/worker-config-drive.md

```diff
@@ -0,0 +1,196 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# worker-config-drive
+
+## Status
+
+provisional
+
+## Table of Contents
+
+<!--ts-->
+   * [worker-config-drive](#worker-config-drive)
+      * [Status](#status)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks [optional]](#drawbacks-optional)
+      * [Alternatives [optional]](#alternatives-optional)
+      * [References](#references)
+
+<!-- Added by: dhellmann, at: 2019-03-14T09:25-0400 -->
+
+<!--te-->
+
+## Summary
+
+Provisioning hosts requires two separate images. The first is the
+primary target image for the host, and contains the operating system
+and other software that the host will run. These are generally
+reusable across many hosts. The second image is the "config drive
+image", which contains configuration settings passed to the target
+image by writing an ISO to a separate partition accessible when the
+host boots. The config drive image is often, but not always,
+customized for each host. It is always often customized for the role a
+host plays (master, worker, etc.).
+
+In order to bring a host into the cluster as a worker we have to give
+it the correct Ignition configuration as it boots.  The target images
+we are using for nodes in an OpenShift (and eventually Kubernetes)
+cluster are RHEL CoreOS images with Ignition configured to look for
+its data in `/openstack/latest` in the config drive. This setting is
+built in to the image, and is not currently configurable.
+
+The Ignition file contents are stored in a Secret within the
+kubernetes database because they contain certificate authority
+information. There are different Ignition settings for worker and
+master nodes, so something will need to know the intended role of each
+host being provisioned in order to select the correct settings.
+
+Something needs to build a config drive image containing the ignition
+file and the baremetal operator needs to pass that formatted image to
+Ironic as part of provisioning the host.
+
+In order to avoid having the baremetal operator coupled tightly to any
+specific images, something that knows about the expectations of the
+target image being deployed (especially the path where the Ignition
+file should be written within the config drive image) should build the
```

Comment:
Is there an example of a config drive that uses a path other than openstack/latest (or openstack/<version>)? My understanding is this is hardcoded.  cloud-init looks [there](https://cloudinit.readthedocs.io/en/latest/topics/datasources/configdrive.html#version-2) (or ec2 for AWS), as does RHCOS openstack images.  For RHCOS baremetal, I thought we discussed on slack it expects to look at a URL instead - does it support config drives at a different path?

Reply:
We've talked about making the RHCOS baremetal image support config drive for the metalkube use case. I would not expect that image to use a path with "openstack" in the name.

Reply:
All of the tooling out there now assumes that a config drive uses the `openstack/` directory.  If you give ironic user data, it uses `openstack/` when it builds the ISO.  Same for gophercloud.    The least friction would be for RHCOS baremetal to follow suit, but it's not ideal to have the word `openstack`, I agree.

 How do we find out what RHCOS baremetal will do? Ironic and/or gophercloud might need changes to support having a different configdrive layout.

Reply:
Likely RHCOS baremetal will do the same as the other images. They will expect the grub config to be changed to set the OEM type. Granted, it is not exactly ideal, and there pretty much no desire on their side to allow for differences. The easiest path is likely going to be to ask for a UEFI openstack image, and just continue to use that.